File: android-java/app/src/main/java/com/oursky/skeleton/ui/MainScreen.java
package com.oursky.skeleton.ui;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.Gravity;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;

import com.oursky.skeleton.R;
import com.oursky.skeleton.client.Login;
import com.oursky.skeleton.helper.LP;
import com.oursky.skeleton.model.MyLoginSession;
import com.oursky.skeleton.redux.AppState;
import com.oursky.skeleton.redux.AppStateObservable;
import com.oursky.skeleton.redux.ClientState;
import com.oursky.skeleton.ui.base.BaseController;
import com.oursky.skeleton.widget.ActionBar;
import com.oursky.skeleton.widget.Button;

import static com.oursky.skeleton.helper.ResourceHelper.dp;
import static com.oursky.skeleton.helper.ResourceHelper.color;
import static com.oursky.skeleton.helper.ResourceHelper.font;

public class MainScreen extends BaseController {
    private ActionBar mActionBar;
    private TextView mMessage;
    private CompositeDisposable mSubscriptions = new CompositeDisposable();

    //region Lifecycle
    //---------------------------------------------------------------
    @Override
    protected @NonNull
    View onCreateView(Context context) {
        LinearLayout contentView = new LinearLayout(context);
        contentView.setOrientation(LinearLayout.VERTICAL);

        mActionBar = new ActionBar(context);
        mActionBar.setTitle(R.string.app_name);
        contentView.addView(mActionBar, LP.linear(LP.MATCH_PARENT, LP.WRAP_CONTENT).build());

        mMessage = new TextView(context);
        mMessage.setTextSize(24);
        mMessage.setTypeface(font(R.font.barlow_condensed_regular));
        contentView.addView(mMessage, LP.linear(LP.WRAP_CONTENT, LP.WRAP_CONTENT).build());

        Button next = new Button(context);
        next.setTextPadding(dp(32), dp(12), dp(32), dp(12));
        next.setBackgroundColor(color(R.color.main_next_bg));
        next.setTextColor(color(R.color.main_next_text));
        next.setTypeface(font(R.font.barlow_condensed_bold));
        next.setTextSize(24);
        next.setText(R.string.main_next);
        contentView.addView(next, LP.linear(LP.MATCH_PARENT, LP.WRAP_CONTENT)
                                    .setMargins(dp(16), dp(16), dp(16), dp(16))
                                    .build());

        // Register event listener
        next.setOnClickListener(onNextClick);
        return contentView;
    }
    @Override
    protected void onAttach(@NonNull View view) {
        super.onAttach(view);
        Observable<AppState> observable = AppStateObservable.create();
        mSubscriptions.add(observable
                .observeOn(AndroidSchedulers.mainThread())
                .map(mapTitle)
                .distinctUntilChanged()
                .subscribe(consumeTitle)
        );
        mSubscriptions.add(observable
                .distinctUntilChanged()
                .observeOn(AndroidSchedulers.mainThread())
                .map(mapLoginState)
                .subscribe(consumeLoginState)
        );
    }
    @Override
    protected void onDetach(@NonNull View view) {
        super.onDetach(view);
        mSubscriptions.clear();
    }
    //---------------------------------------------------------------
    //endregion

    //region UI Events
    //---------------------------------------------------------------
    private final View.OnClickListener onNextClick = (view) -> {
        pushController(new SecondScreen());
    };
    //---------------------------------------------------------------
    //endregion

    //region Redux
    //---------------------------------------------------------------
    private final Function<AppState,String> mapTitle = (state) -> state.view().title;
    private final Consumer<String> consumeTitle = (title) -> {
        mActionBar.setTitle(title);
    };
    private final Function<AppState,ClientState.APIState<Login.Output>>
            mapLoginState = (state) -> state.client().login;
    private final Consumer<ClientState.APIState<Login.Output>> consumeLoginState = (mapped) -> {
        if (mapped.data == null || mapped.data.me == null) return;
        mMessage.setText(getResources().getString(R.string.main_message, mapped.data.me.name));
    };
    //---------------------------------------------------------------
    //endregion
}

File: android-kotlin/app/src/main/java/com/oursky/skeleton/ui/MainScreen.kt
package com.oursky.skeleton.ui

import android.content.Context
import android.view.View
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.Consumer
import io.reactivex.functions.Function
import com.oursky.skeleton.R
import com.oursky.skeleton.helper.LP
import com.oursky.skeleton.helper.ResourceHelper.color
import com.oursky.skeleton.helper.ResourceHelper.dp
import com.oursky.skeleton.helper.ResourceHelper.font
import com.oursky.skeleton.client.Login
import com.oursky.skeleton.redux.ClientStore
import com.oursky.skeleton.redux.ViewStore
import com.oursky.skeleton.ui.base.BaseController
import com.oursky.skeleton.widget.ActionBar
import com.oursky.skeleton.widget.Button
import com.oursky.skeleton.MainApplication.Companion.store

class MainScreen : BaseController() {
    private val mSubscriptions = CompositeDisposable()
    private var mActionBar: ActionBar? = null
    private var mMessage: TextView? = null

    //region Lifecycle
    //---------------------------------------------------------------
    override fun onCreateView(context: Context): View {
        val contentView = LinearLayout(context)
        contentView.orientation = LinearLayout.VERTICAL

        mActionBar = ActionBar(context)
        mActionBar?.setTitle(R.string.login_title)
        contentView.addView(mActionBar, LP.linear(LP.MATCH_PARENT, LP.WRAP_CONTENT).build())

        mMessage = EditText(context)
        mMessage?.textSize = 20f
        mMessage?.typeface = font(R.font.barlow_condensed_regular)
        mMessage?.setHint(R.string.login_name_hint)
        contentView.addView(mMessage, LP.linear(LP.MATCH_PARENT, LP.WRAP_CONTENT).build())

        val next = Button(context)
        next.setTextPadding(dp(32), dp(12), dp(32), dp(12))
        next.setBackgroundColor(color(R.color.main_next_bg))
        next.setTextColor(color(R.color.main_next_text))
        next.setTypeface(font(R.font.barlow_condensed_bold))
        next.setTextSize(24f)
        next.setText(R.string.main_next)
        contentView.addView(next, LP.linear(LP.MATCH_PARENT, LP.WRAP_CONTENT)
                                    .setMargins(dp(16), dp(16), dp(16), dp(16))
                                    .build())
        // Register event listener
        next.setOnClickListener(onNextClick)
        return contentView
    }
    override fun onAttach(view: View) {
        super.onAttach(view)
        mSubscriptions.add(store().observe(store().view)
                .distinctUntilChanged()
                .observeOn(AndroidSchedulers.mainThread())
                .map(mapTitle)
                .subscribe(consumeTitle)
        )
        mSubscriptions.add(store().observe(store().client)
                .distinctUntilChanged()
                .observeOn(AndroidSchedulers.mainThread())
                .map(mapLoginState)
                .subscribe(consumeLoginState)
        )
    }
    override fun onDetach(view: View) {
        super.onDetach(view)
        mSubscriptions.clear()
    }
    //---------------------------------------------------------------
    //endregion

    //region UI Events
    //---------------------------------------------------------------
    private val onNextClick = View.OnClickListener { _: View ->
        pushController(SecondScreen())
    }
    //---------------------------------------------------------------
    //endregion

    //region Redux
    //---------------------------------------------------------------
    private val mapTitle = Function<ViewStore.State, String> { state ->
        state.title
    }
    private val consumeTitle = Consumer<String> { mapped ->
        mActionBar?.setTitle(mapped)
    }
    private val mapLoginState = Function<ClientStore.State, ClientStore.APIState<Login.Output>> { state ->
        state.login
    }
    private val consumeLoginState = Consumer<ClientStore.APIState<Login.Output>> { mapped ->
        mapped.data?.me?.let {
            mMessage?.text = resources?.getString(R.string.main_message, it.name)
        }
    }
    //---------------------------------------------------------------
    //endregion
}
